sql: allow stars inside view definitions

knz opened the pull request:
First commit from #97544.
Fixes #10028.

Release note (sql change): It is now possible to use `*` inside CREATE VIEW. The list of columns is expanded at the time the view is created, so that new columns added after the view was defined are not included in the view. This behavior is the same as PostgreSQL.

knz committed:
sql: allow stars inside view definitions

Release note (sql change): It is now possible to use `*` inside CREATE
VIEW. The list of columns is expanded at the time the view is created,
so that new columns added after the view was defined are not included
in the view. This behavior is the same as PostgreSQL.

knz commented on pkg/sql/logictest/testdata/logic_test/views:
```diff
@@ -380,68 +380,170 @@ CREATE VIEW virt2 AS SELECT range_id, lease_holder FROM crdb_internal.ranges
+        ) AS SELECT t.a, t.b FROM test.public.t
+star3   CREATE VIEW public.star3 (
+          a
+        ) AS SELECT a FROM test.public.t ORDER BY t.*
```
I am noting how the stars inside ORDER BY and GROUP BY are preserved here. This means they are also perserved in UDFs.

I believe this is OK? But I think we need some semantic expert to double check. In any case, this choice is not specific to this PR so it should not block it.

DrewKimball reviewed (commented):
> Reviewed 1 of 4 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/97515)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @knz, @mgartner, and @rharding6373)

___
*[`pkg/sql/opt/optbuilder/project.go` line 151 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/97515#-NOuyF-M3AR9y9t8CNw-:-NOuyF-M3AR9y9t8CNw0:b-csdgmr) ([raw file](https://github.com/cockroachdb/cockroach/blob/09d2c5df77c9fef1b0fb6bb277c54e72facbca78/pkg/sql/opt/optbuilder/project.go#L151)):*
> ```Go
> 
> 					aliases, exprs := b.expandStar(e.Expr, inScope)
> 					if b.insideFuncDef || b.insideViewDef {
> ```

I think there's another case in `Builder.buildDataSource` that needs to be changed (see #96375).
___
*[`pkg/sql/logictest/testdata/logic_test/views` line 439 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/97515#-NOuzigO0_wLa00CewJt:-NOuzigO0_wLa00CewJu:b-jwituc) ([raw file](https://github.com/cockroachdb/cockroach/blob/09d2c5df77c9fef1b0fb6bb277c54e72facbca78/pkg/sql/logictest/testdata/logic_test/views#L439)):*
> ```
> 
> statement ok
> CREATE VIEW star16 AS SELECT t1.a, t2.a AS a2 FROM t AS t1 JOIN t AS t2 ON t1.a IN (SELECT a FROM (SELECT * FROM t))
> ```

Could you also add the cases from #96375? (similar to the existing join test cases, but without aliases)
___
*[`pkg/sql/logictest/testdata/logic_test/views` line 486 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/97515#-NOus2K3iUKFUQkqNrPs-r1-486:-NOv-RTqCh4V83XhtvW6:b81o75) ([raw file](https://github.com/cockroachdb/cockroach/blob/09d2c5df77c9fef1b0fb6bb277c54e72facbca78/pkg/sql/logictest/testdata/logic_test/views#L486)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

I am noting how the stars inside ORDER BY and GROUP BY are preserved here. This means they are also perserved in UDFs.

I believe this is OK? But I think we need some semantic expert to double check. In any case, this choice is not specific to this PR so it should not block it.
</blockquote></details>

That's a good catch, and I think it's a bug.


<!-- Sent from Reviewable.io -->

knz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/97515)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @DrewKimball, @mgartner, and @rharding6373)

___
*[`pkg/sql/opt/optbuilder/project.go` line 151 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/97515#-NOuyF-M3AR9y9t8CNw-:-NOv1R3n7IFJ-p9bam8R:bkhzaye) ([raw file](https://github.com/cockroachdb/cockroach/blob/09d2c5df77c9fef1b0fb6bb277c54e72facbca78/pkg/sql/opt/optbuilder/project.go#L151)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

I think there's another case in `Builder.buildDataSource` that needs to be changed (see #96375).
</blockquote></details>

hum no that would be a regression - we currently allow views to reference subqueries withut an alias, so removing that here would be a regression from current behavior.
___
*[`pkg/sql/logictest/testdata/logic_test/views` line 439 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/97515#-NOuzigO0_wLa00CewJt:-NOv1jjB-SrsvVBUqZX9:b-zii22h) ([raw file](https://github.com/cockroachdb/cockroach/blob/09d2c5df77c9fef1b0fb6bb277c54e72facbca78/pkg/sql/logictest/testdata/logic_test/views#L439)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Could you also add the cases from #96375? (similar to the existing join test cases, but without aliases)
</blockquote></details>

i'll look into it.
___
*[`pkg/sql/logictest/testdata/logic_test/views` line 486 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/97515#-NOus2K3iUKFUQkqNrPs-r1-486:-NOv1hmtDUoppqbGqPv7:b1p62rl) ([raw file](https://github.com/cockroachdb/cockroach/blob/09d2c5df77c9fef1b0fb6bb277c54e72facbca78/pkg/sql/logictest/testdata/logic_test/views#L486)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

That's a good catch, and I think it's a bug.
</blockquote></details>

i'll file an issue.


<!-- Sent from Reviewable.io -->

knz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/97515)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @DrewKimball, @mgartner, and @rharding6373)

___
*[`pkg/sql/logictest/testdata/logic_test/views` line 486 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/97515#-NOus2K3iUKFUQkqNrPs-r1-486:-NOv2zlD2GywOB4dQzHj:b8v3wkp) ([raw file](https://github.com/cockroachdb/cockroach/blob/09d2c5df77c9fef1b0fb6bb277c54e72facbca78/pkg/sql/logictest/testdata/logic_test/views#L486)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

i'll file an issue.
</blockquote></details>

#97520 


<!-- Sent from Reviewable.io -->

knz reviewed (commented):
RFAL

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/97515#-:-NOv54Sp7OtlYuR-LySf:bb72pd0)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @DrewKimball, @mgartner, and @rharding6373)

___
*[`pkg/sql/opt/optbuilder/project.go` line 151 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/97515#-NOuyF-M3AR9y9t8CNw-:-NOv4skHEIJUaWIEvbkb:b-auk9rp) ([raw file](https://github.com/cockroachdb/cockroach/blob/09d2c5df77c9fef1b0fb6bb277c54e72facbca78/pkg/sql/opt/optbuilder/project.go#L151)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

hum no that would be a regression - we currently allow views to reference subqueries withut an alias, so removing that here would be a regression from current behavior.
</blockquote></details>

I double checked - we do not need the special case related to #96375 here because CREATE VIEW separately rejects duplicate column names, so if there's an ambiguity it will be prevented this way.
___
*[`pkg/sql/logictest/testdata/logic_test/views` line 439 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/97515#-NOuzigO0_wLa00CewJt:-NOv4rti6KA4xIQCQWpC:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/09d2c5df77c9fef1b0fb6bb277c54e72facbca78/pkg/sql/logictest/testdata/logic_test/views#L439)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

i'll look into it.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

rharding6373 reviewed (approved):
Very nice, thanks for all the test cases! I only have suggestions for a couple more. <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/97515#-:-NOv9jsI5N13kx585enJ:bvpogh5)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajwerner, @DrewKimball, @knz, and @mgartner)

___
*[`pkg/sql/logictest/testdata/logic_test/views` line 440 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/97515#-NOv7tcnBgAQaPDsMmjx:-NOv7tcnBgAQaPDsMmjy:b-wt6nr3) ([raw file](https://github.com/cockroachdb/cockroach/blob/09d2c5df77c9fef1b0fb6bb277c54e72facbca78/pkg/sql/logictest/testdata/logic_test/views#L440)):*
> ```
> statement ok
> CREATE VIEW star16 AS SELECT t1.a, t2.a AS a2 FROM t AS t1 JOIN t AS t2 ON t1.a IN (SELECT a FROM (SELECT * FROM t))
> ```

Could you please add the following tests (or similar) to make sure schema changes and dependencies are handled correctly? 

```
statement ok
ALTER TABLE t ADD COLUMN c INT;

statement error
ALTER TABLE t DROP COLUMN b;

# we expect the columns in the view to be unchanged
statement ok
ALTER TABLE t RENAME COLUMN b TO d;
```


<!-- Sent from Reviewable.io -->

DrewKimball reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/97515#-:-NOvBk3QCjJhvGICi-xf:bnfp4nl)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @ajwerner, @knz, and @mgartner)

___
*[`pkg/sql/opt/optbuilder/project.go` line 151 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/97515#-NOuyF-M3AR9y9t8CNw-:-NOv77wX6CCcjWtDjz00:ba7duo0) ([raw file](https://github.com/cockroachdb/cockroach/blob/09d2c5df77c9fef1b0fb6bb277c54e72facbca78/pkg/sql/opt/optbuilder/project.go#L151)):*
>CREATE VIEW separately rejects duplicate column names

I think this will only apply to the top-level columns returned by the view, but it seems ok anyway since apparently views bind differently from UDFs: https://github.com/cockroachdb/cockroach/issues/96375#issuecomment-1440882770
___
*[`pkg/sql/logictest/testdata/logic_test/views` line 468 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/97515#-NOv7Z5EA3MoUzLxAGWl:-NOv7Z5EA3MoUzLxAGWm:b-17u3rs) ([raw file](https://github.com/cockroachdb/cockroach/blob/8298d85cc8ab00ea363c7d3cc8ea82941db89deb/pkg/sql/logictest/testdata/logic_test/views#L468)):*
> ```
> 
> 
> # This error is the same as in postgres.
> ```

[nit] postgres gives `subquery in FROM must have an alias`, so not the same error here.


<!-- Sent from Reviewable.io -->

knz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/97515)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @ajwerner, @DrewKimball, @mgartner, and @rharding6373)

___
*[`pkg/sql/opt/optbuilder/project.go` line 151 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/97515#-NOuyF-M3AR9y9t8CNw-:-NOwMtp8En3uHFbNgUFd:b3tcbt7) ([raw file](https://github.com/cockroachdb/cockroach/blob/09d2c5df77c9fef1b0fb6bb277c54e72facbca78/pkg/sql/opt/optbuilder/project.go#L151)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

>CREATE VIEW separately rejects duplicate column names

I think this will only apply to the top-level columns returned by the view, but it seems ok anyway since apparently views bind differently from UDFs: https://github.com/cockroachdb/cockroach/issues/96375#issuecomment-1440882770
</blockquote></details>

I investigated this further and we have indeed an 'interesting' situation here:
```sql
select array[z.*]::string from (select * from (select a from t), (select a from t)) z;
```

This example is interesting because there's just 1 unambiguous "outer" column, but we need to be careful "in the middle".

With my current code, we get invalid behavior as follows:
```
demo@127.0.0.1:26257/movr> show create view v;
  table_name |                                       create_statement
-------------+-----------------------------------------------------------------------------------------------
  v          | CREATE VIEW public.v (
             |     "array"
             | ) AS SELECT
             |         ARRAY[z.*]::STRING
             |     FROM
             |         (SELECT a, a FROM (SELECT a FROM movr.public.t), (SELECT a FROM movr.public.t)) AS z
(1 row)

Time: 9ms total (execution 9ms / network 0ms)

demo@127.0.0.1:26257/movr> select * from v;
ERROR: column reference "a" is ambiguous (candidates: <anonymous>.a)
SQLSTATE: 42702
```

Which means my PR needs to solve a subset of #96375 before we consider the job done.
___
*[`pkg/sql/logictest/testdata/logic_test/views` line 440 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/97515#-NOv7tcnBgAQaPDsMmjx:-NOwK2wc9IOpe1Kd2eWR:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/09d2c5df77c9fef1b0fb6bb277c54e72facbca78/pkg/sql/logictest/testdata/logic_test/views#L440)):*
<details><summary><i>Previously, rharding6373 (Rachael Harding) wrote…</i></summary><blockquote>

Could you please add the following tests (or similar) to make sure schema changes and dependencies are handled correctly? 

```
statement ok
ALTER TABLE t ADD COLUMN c INT;

statement error
ALTER TABLE t DROP COLUMN b;

# we expect the columns in the view to be unchanged
statement ok
ALTER TABLE t RENAME COLUMN b TO d;
```
</blockquote></details>

Done.
___
*[`pkg/sql/logictest/testdata/logic_test/views` line 468 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/97515#-NOv7Z5EA3MoUzLxAGWl:-NOwKenI93DHNf9A0sRP:bklp7k5) ([raw file](https://github.com/cockroachdb/cockroach/blob/8298d85cc8ab00ea363c7d3cc8ea82941db89deb/pkg/sql/logictest/testdata/logic_test/views#L468)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

[nit] postgres gives `subquery in FROM must have an alias`, so not the same error here.
</blockquote></details>

Good point. Changed to add a test using pg's syntax.


<!-- Sent from Reviewable.io -->

knz committed:
sql/opt: support anonymous subqueries in CREATE FUNCTION

Release note (bug fix): CockroachDB now supports using subqueries in
UDFs without an AS clause, for consistency with the syntax supported
outside of UDFs. (Subqueries without an AS clause is a CRDB-specific
extension.)

knz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/97515)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @ajwerner, @DrewKimball, @mgartner, and @rharding6373)

___
*[`pkg/sql/opt/optbuilder/project.go` line 151 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/97515#-NOuyF-M3AR9y9t8CNw-:-NOwVwUwEkylYOYgQQrk:b-d1jtsr) ([raw file](https://github.com/cockroachdb/cockroach/blob/09d2c5df77c9fef1b0fb6bb277c54e72facbca78/pkg/sql/opt/optbuilder/project.go#L151)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

I investigated this further and we have indeed an 'interesting' situation here:
```sql
select array[z.*]::string from (select * from (select a from t), (select a from t)) z;
```

This example is interesting because there's just 1 unambiguous "outer" column, but we need to be careful "in the middle".

With my current code, we get invalid behavior as follows:
```
demo@127.0.0.1:26257/movr> show create view v;
  table_name |                                       create_statement
-------------+-----------------------------------------------------------------------------------------------
  v          | CREATE VIEW public.v (
             |     "array"
             | ) AS SELECT
             |         ARRAY[z.*]::STRING
             |     FROM
             |         (SELECT a, a FROM (SELECT a FROM movr.public.t), (SELECT a FROM movr.public.t)) AS z
(1 row)

Time: 9ms total (execution 9ms / network 0ms)

demo@127.0.0.1:26257/movr> select * from v;
ERROR: column reference "a" is ambiguous (candidates: <anonymous>.a)
SQLSTATE: 42702
```

Which means my PR needs to solve a subset of #96375 before we consider the job done.
</blockquote></details>

Rebased on top of #97544, which fixes this. PTAL.


<!-- Sent from Reviewable.io -->

knz commented:
TFYRs!

bors r=DrewKimball,rharding6373
